core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -0,0 +1,279 @@
+
+      val path: String = toResourcePath(resource)
+      ZkUtils.updatePersistentPath(zkClient, path, Json.encode(Acl.toJsonCompatibleMap(filteredAcls)))
+
+      updateAclChangedFlag(resource)
+
+      true
+    } else false
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    if (ZkUtils.pathExists(zkClient, toResourcePath(resource))) {
+      ZkUtils.deletePath(zkClient, toResourcePath(resource))
+      updateAclChangedFlag(resource)
+      true
+    } else false
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      aclCache.get(resource).getOrElse(Set.empty[Acl])
+    }
+  }
+
+  private def getAclsFromZk(resource: Resource): Set[Acl] = {
+    val aclJson = ZkUtils.readDataMaybeNull(zkClient, toResourcePath(resource))._1
+    aclJson.map(Acl.fromJson).getOrElse(Set.empty)
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Set[Acl] = {
+    val aclSets = aclCache.values
+    var acls = Set.empty[Acl]
+    for (aclSet <- aclSets) {
+      acls ++= aclSet.filter(acl => acl.principal == principal)
+    }
+    acls
+  }
+
+  private def loadCache: Set[Acl] = {
+    var acls = Set.empty[Acl]
+    val resourceTypes = ZkUtils.getChildren(zkClient, SimpleAclAuthorizer.AclZkPath)
+    for (rType <- resourceTypes) {
+      val resourceType = ResourceType.fromString(rType)
+      val resourceTypePath = SimpleAclAuthorizer.AclZkPath + "/" + resourceType.name
+      val resourceNames = ZkUtils.getChildren(zkClient, resourceTypePath)
+      for (resourceName <- resourceNames)
+        acls ++= getAclsFromZk(Resource(resourceType, resourceName.toString))
+    }
+    acls
+  }
+
+
+  private def toResourcePath(resource: Resource): String = {
+    SimpleAclAuthorizer.AclZkPath + "/" + resource.resourceType + "/" + resource.name
+  }
+
+  private def logAuditMessage(principal: KafkaPrincipal, authorized: Boolean, operation: Operation, resource: Resource, host: String): Unit = {
+    val permissionType = if (authorized) "Allowed" else "Denied"
+    authorizerLogger.debug(s"Principal = $principal is $permissionType Operation = $operation from host = $host on resource = $resource")
+  }
+
+  private def updateAclChangedFlag(resource: Resource): Unit = {
+    ZkUtils.createSequentialPersistentPath(zkClient, SimpleAclAuthorizer.AclChangedZkPath + "/" + SimpleAclAuthorizer.AclChangedPrefix, resource.toString)
+  }
+
+  object AclChangedNotificaitonHandler extends NotificationHandler {
+
+    override def processNotification(notificationMessage: Option[String]): Unit = {
+      val resource: Resource = Resource.fromString(notificationMessage.get)
+      val acls = getAclsFromZk(resource)
+
+      inWriteLock(lock) {
+        if (acls.nonEmpty)
+          aclCache.put(resource, acls)
+        else
+          aclCache.remove(resource)
+      }
+    }
+  }
+
+  object ZkStateChangeListener extends IZkStateListener {
+
+    var reconnection = false
```

Comment:
In the initialize we register this zkClient so in case a reconnection happens zkClient will invoke this method and as part handling a new session it will process any missed notifications during the reconnection.